src/components/rotate-object-button.js

```diff
@@ -0,0 +1,334 @@
+    this.el.removeEventListener("grab-end", this.onGrabEnd);
+  }
+});
+
+AFRAME.registerSystem("rotate-selected-object", {
+  init() {
+    this.o = null;
+    this.mode = null;
+    this.active = false;
+
+    this.puppet = {
+      Ci: new THREE.Quaternion(),
+      Ci_inverse: new THREE.Quaternion(),
+      Cc: new THREE.Quaternion(),
+      Cd: new THREE.Quaternion(),
+      Oi: new THREE.Quaternion(),
+      Oc: new THREE.Quaternion(),
+      CcSnapper: new THREE.Euler(),
+      snappedCc: new THREE.Quaternion()
+    };
+
+    this.planarInfo = {
+      plane: new THREE.Mesh(
+        new THREE.PlaneBufferGeometry(100000, 100000, 2, 2),
+        new THREE.MeshBasicMaterial({
+          visible: true,
+          wireframe: true,
+          side: THREE.DoubleSide,
+          transparent: true,
+          opacity: 0.3
+        })
+      ),
+      normal: new THREE.Vector3(),
+      intersections: [],
+      Pi: new THREE.Vector3(),
+      Pp: new THREE.Vector3(),
+      Pc: new THREE.Vector3(),
+      PiPc: new THREE.Vector3(),
+      PpPc: new THREE.Vector3(),
+      XYi: new THREE.Vector3()
+    };
+
+    this.right = new THREE.Vector3(1, 0, 0);
+    this.up = new THREE.Vector3(0, 1, 0);
+    this.forward = new THREE.Vector3(0, 0, 1);
+    this.v = new THREE.Vector3();
+    this.v2 = new THREE.Vector3();
+    this.q = new THREE.Quaternion();
+    this.q2 = new THREE.Quaternion();
+
+    this.dzAll = 0;
+    this.dzStore = 0;
+    this.dzApplied = 0;
+    this.dyAll = 0;
+    this.dyStore = 0;
+    this.dyApplied = 0;
+
+    this.resetInfo = {
+      obj: new THREE.Vector3(),
+      eye: new THREE.Vector3(),
+      eyeToObj: new THREE.Vector3(),
+      resetTarget: new THREE.Vector3()
+    };
+    this.cameraWorldQuaternion = new THREE.Quaternion();
+    this.objectWorldQuaternion = new THREE.Quaternion();
+
+    this.el.object3D.add(this.planarInfo.plane);
+  },
+
+  rotate(o, hand, data) {
+    this.o = o;
+    this.hand = hand.id === "cursor" ? document.querySelector("#player-right-controller").object3D : hand.object3D;
+    this.mode = data.mode;
+    this.camera = this.camera || document.querySelector("#player-camera").object3D;
+    if (this.mode === ROTATE_MODE.AXIS) {
+      this.axis.copy(data.axis);
+    }
+    if (this.mode === ROTATE_MODE.AXIS || this.mode === ROTATE_MODE.FREE || this.mode === ROTATE_MODE.GARY) {
+      const { plane, intersections, Pi, Pp } = this.planarInfo;
+      const v = this.v;
+      const q = this.q;
+
+      this.camera.matrixWorld.decompose(v, plane.quaternion, v);
+      this.o.matrixWorld.decompose(plane.position, q, v);
+      plane.matrixNeedsUpdate = true;
+      plane.updateMatrixWorld(true);
+
+      this.raycaster =
+        this.raycaster || document.querySelector("#cursor-controller").components["cursor-controller"].raycaster;
+      intersections.length = 0;
+      const far = this.raycaster.far;
+      this.raycaster.far = 1000;
+      plane.raycast(this.raycaster, intersections);
```

Comment:
postponing for now because i like being able to turn on the plane while debugging and also i don't know the point of the "constant" parameter of the three.js Plane object